iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java

```diff
@@ -0,0 +1,232 @@
+    for (int tryCount = 0; tryCount < clientSize; ++tryCount) {
+      final int clientIndex = (int) (currentClientIndex++ % clientSize);
+      Pair<IoTDBThriftSyncConnectorClient, Boolean> clientAndStatus =
+          endPoint2client.get(endPoints.get(clientIndex));
+      if (Boolean.TRUE.equals(clientAndStatus.getRight())) {
+        return clientAndStatus;
+      }
+    }
+    throw new PipeConnectionException(
+        "All clients are dead, please check the connection to the receiver.");
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeInsertNodeTabletInsertionEvent event) {
+    TEndPoint endPoint = leaderCacheManager.parseEndPoint(event);
+    return endPoint == null
+            || !useLeaderCache
+            || !endPoint2client.containsKey(endPoint)
+            || Boolean.FALSE.equals(endPoint2client.get(endPoint).getRight())
+        ? getOneConnectedClientAndStatus()
+        : endPoint2client.get(endPoint);
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeRawTabletInsertionEvent event) {
+    TEndPoint endPoint = leaderCacheManager.parseEndPoint(event);
+    return endPoint == null
+            || !useLeaderCache
+            || !endPoint2client.containsKey(endPoint)
+            || Boolean.FALSE.equals(endPoint2client.get(endPoint).getRight())
+        ? getOneConnectedClientAndStatus()
+        : endPoint2client.get(endPoint);
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeTsFileInsertionEvent event) {
+    return getOneConnectedClientAndStatus();
+  }
+
+  public void updateOrCreate(String deviceId, TEndPoint endPoint) {
+    try {
+      endPoints.add(endPoint);
+      initAndSetClient(endPoint);
+      leaderCacheManager.updateOrCreate(deviceId, endPoint);
+    } catch (TTransportException e) {
+      LOGGER.warn(
+          "Unable to create client with target server ip: {}, port: {}, because: {}.",
+          endPoint.getIp(),
+          endPoint.getPort(),
+          e.getMessage());
+    } catch (IOException e) {
+      LOGGER.warn("Unable to create a handshake request.");
+    }
+  }
+
+  public void clean() {
+    for (TEndPoint endPoint : endPoint2client.keySet()) {
+      try {
+        if (endPoint2client.get(endPoint).getLeft() != null) {
+          endPoint2client.get(endPoint).getLeft().close();
+          endPoint2client.get(endPoint).setLeft(null);
+          endPoint2client.get(endPoint).setRight(null);
+        }
+      } catch (Exception e) {
+        LOGGER.warn(
+            "Failed to close client {}:{}, because: {}.", endPoint.getIp(), endPoint.getPort(), e);
+      } finally {
+        endPoint2client.get(endPoint).setRight(false);
+      }
+    }
+  }
+
+  private void initAndSetClient(TEndPoint endPoint) throws TTransportException, IOException {
+    if (endPoint2client.get(endPoint).getLeft() != null) {
```

Comment:
Closing a client upon redirection is a little bit not graceful...... Besides, if a client is killed here, the status will not be set to false, and if it does not start successfully....